espressostreamer/espresso_streamer.go

```diff
@@ -168,36 +158,23 @@ func (s *EspressoStreamer) Advance() {
 // It is a do-while loop, which means it will always execute at least once.
 //
 // Expose the *parseHotShotPayloadFn* to the caller for testing purposes
-func (s *EspressoStreamer) QueueMessagesFromHotShotUntil(
+func (s *EspressoStreamer) QueueMessagesFromHotshot(
 	ctx context.Context,
 	parseHotShotPayloadFn func(tx espressoTypes.Bytes) ([]*MessageWithMetadataAndPos, error),
-	condition func(messages []*MessageWithMetadataAndPos) bool,
 ) error {
-	for {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		default:
-			messages, err := fetchNextHotshotBlock(ctx, s.espressoClient, s.nextHotshotBlockNum, parseHotShotPayloadFn, s.namespace)
-			if err != nil {
-				// TODO: minimize the log output
-				log.Error("failed to fetch next hotshot block", "err", err)
-				time.Sleep(s.retryTime)
-				continue
-			}
-
-			if len(messages) > 0 {
-				s.messageWithMetadataAndPos = append(s.messageWithMetadataAndPos, messages...)
-			}
-			s.nextHotshotBlockNum += 1
-
-			if condition(messages) {
-				return nil
-			}
-
-			time.Sleep(s.hotshotPollingInterval)
-		}
+	s.messageLock.Lock()
+	defer s.messageLock.Unlock()
+
+	messages, err := fetchNextHotshotBlock(ctx, s.espressoClient, s.nextHotshotBlockNum, parseHotShotPayloadFn, s.namespace)
+	if err != nil {
+		return err
 	}
+
+	if len(messages) > 0 {
+		s.messageWithMetadataAndPos = append(s.messageWithMetadataAndPos, messages...)
+	}
+	s.nextHotshotBlockNum += 1
```

Comment:
should we increase the nextHotshotBlockNum even if there is an error?

Reply:
We should not skip any hotshot blocks. And if errors occur, that's probably network issues. If not, we should  stop and check our code instead of continue consuming I think.